modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/prepare/PrepareServiceImpl.java

```diff
@@ -238,26 +241,102 @@ public CompletableFuture<QueryPlan> prepareAsync(
                 .parameters(Commons.arrayToMap(operationContext.parameters()))
                 .build();
 
-        result = prepareAsync0(parsedResult, planningContext);
+        QueryTransactionContext txContext = operationContext.txContext();
+
+        // Validate statement
+        CompletableFuture<ValidStatement> validFut = validateStatement(parsedResult, planningContext);
+
+        // Try optimize to fast
+        CompletableFuture<FastOptimizationResult> plannedOrValidated =
```

Comment:
tbh, I didn't get the idea behind introduction of additional container for composite result.

it's better to do this:
```
        CompletableFuture<ValidStatement> validFut = validateStatement(parsedResult, planningContext);

        result = validFut.thenCompose(stmt -> {
            QueryPlan plan = tryOptimizeFast(stmt, planningContext, txContext);

            // If optimize fast returned a plan, return it.
            if (plan != null) {
                return CompletableFuture.completedFuture(plan);
            }

            // Otherwise, continue with the regular planning.
            return prepareAsync0(stmt, planningContext, txContext);
        });
```

Reply:
Fixed.